homeassistant/components/rotel/media_player.py

```diff
@@ -0,0 +1,493 @@
+        if self._conn is not None:
+            self._conn.close()
+
+    def start_read_messages(self, hass):
+        """Create a task to start reading messages."""
+        self._read_messages_task = hass.loop.create_task(
+            self.async_read_messages()
+        )
+
+    async def cancel_read_messages(self):
+        """Cancel the _read_messages_task."""
+        if self._read_messages_task is not None:
+            _LOGGER.info("Cancelling read_messages task.  Done was: %r.",
+                         self._read_messages_task.done())
+            if not self._read_messages_task.done():
+                self._read_messages_task.cancel()
+                await self._read_messages_task
+            else:
+                ex = self._read_messages_task.exception()
+                if ex is not None:
+                    _LOGGER.error(
+                        "Read messages task contained an exception.",
+                        exc_info=ex)
+            self._read_messages_task = None
+
+    async def async_read_messages(self):
+        """
+        Read messages published by the device.
+
+        Read messages published by the device and use them to maintain the
+        state of this object.
+        Schedule a DISPLAY_REFRESH command before we start reading.
+        If the device is already on then this is a null command that will
+        simply trigger a feedback message that will sync the state of this
+        object with the physical device.
+        """
+        _LOGGER.info("Message reader started for %s.", self.name)
+        try:
+            await self.send_command('DISPLAY_REFRESH')
+            async for message in self._conn.read_messages():
+                _LOGGER.debug("Message received by %s.", self.name)
+                self.handle_message(message)
+        except asyncio.CancelledError:
+            _LOGGER.info("Message reader cancelled for %s", self.name)
+        except Exception:
+            _LOGGER.error((
+                "Message reader for '%s' exiting due to unexpected exception. "
+                "Message reader can be reinstated by calling '%s' service."
+            ), self.name, SERVICE_RECONNECT)
+            raise
+
+    async def reconnect(self):
+        """
+        Reconnect.
+
+        Don't call this more often than needed - it is asynchronous
+        and it doesn't check for activity in progress on the connection
+        so there could be a risk of a conflict.
+        """
+        await self.cancel_read_messages()
+
+        # Ignore any errors while closing the connection because
+        # the reason we'd be doing this would probably be due to some
+        # sort of issue with the existing connection anyway.
+        try:
+            self.close_connection()
+        # pylint: disable=broad-except
+        except Exception:
+            _LOGGER.error("Could not close connection", exc_info=True)
+            _LOGGER.warning("Ignoring error and attempting to reconnect")
```

Comment:
Error means something failed and the operation failed.  Warning means something is wrong, but the operation can continue.  Log one or the other.